Migration page from "old" notifiers to "new" notifiers (#2793)
rrousselGit/riverpod: website/docs/migration/from_state_notifier.mdx

---
title: From `StateNotifier` and `ChangeNotifier`
---

import buildIinit from "./build_init";
import familyAndDispose from "./family_and_dispose";
import asyncNotifier from "./async_notifier";
import addListener from "./add_listener";
import fromChangeNotifier from "./from_change_notifier";
import fromStateProvider from "./from_state_provider";
import oldLifecycles from "./old_lifecycles";
import { AutoSnippet } from "../../src/components/CodeSnippet";

> Riverpod went `2.0`! To see the full list of changes, consult the
> [Changelog](https://pub.dev/packages/flutter_riverpod/changelog#200).
> 
> Riverpod `2.0` introduced the new `Notifier` and `AsyncNotifier` classes.  

Reviewer: Not a fan of this introduction. Here's my attempt :) 
```suggestion
Along with Riverpod [2.0](https://pub.dev/packages/flutter_riverpod/changelog#200), new classes were introduced: `Notifier`/`AsyncNotifier`.
The good old `StateNotifier`/`ChangeNotifier` classes are now discouraged in favour of those new APIs.
```

This page shows how to migrate from the deprecated `StateNotifier` (and from the legacy `ChangeNotifier`)
to the new APIs.

The main benefit introduced by `AsyncNotifier` is a better `async` support; indeed,

> `AsyncNotifier` can be thought as a `FutureProvider` with methods.

Reviewer: "methods" may be unclear

```suggestion
`AsyncNotifier` can be thought as a `FutureProvider` which can expose ways to be modified from the UI.
```

> Furthermore, the new `Notifier`s:

Reviewer: ```suggestion
Furthermore, the new `Async/Notifier`s:
```

- Expose a `Ref` object inside its class
- Offer similar syntax between codegen and non-codegen approaches
- Offer similar syntax between their sync and async versions
- Move away logic from Providers and centralize it into the Notifiers themselves

Let's see how to define a `Notifier`, how it compares with `StateNotifier` and explore
the new `AsyncNotifier` for asynchronous state classes.

> ## The new syntax

Reviewer: This page should be littered with links to the Performing Side effects page
https://deploy-preview-2740--river-pod.netlify.app/docs/essentials/side_effects

It already extensively explains the syntax <img width="973" alt="Screenshot 2023-08-29 at 08 51 36" src="https://github.com/rrousselGit/riverpod/assets/20165741/114eb963-a15b-416c-a412-5b4d18a82d43">

> ## The new syntax

Reviewer: This doesn't talk about how to invoke notifier methods in widgets.

Although it's the same for both, that's an important note.

Let's write some simple logic with a synchronous `Notifier`:

> <AutoSnippet language="dart" {...buildIinit}></AutoSnippet>

Reviewer: It feels backward to show Notifier before StateNotifier

Notice the following design choices:

- Notifiers are initialized in their `build` method, along with their reactive dependencies

> - Notifiers expose `ref` into their class, whereas their Provider don't

Reviewer: Use a lower case when talking about providers as a kind.

Provider is for the class `Provider`

```suggestion
- Notifiers expose `ref` into their class, whereas their provider don't
```

### Comparison with `StateNotifier`

Let's compare the new APIs with the old ones; here, we implement the above example with `StateNotifier`:

> ```dart

Reviewer: No code blocks in docs please :)

Reviewer: You don't need to specify hooks/codegen objects. Only the "raw" parameter is required

> class WellNotifier extends StateNotifier<int> {

Reviewer: In the end, it's a counter right? :D
I think making a "normal" Counter would probably be more readable. Everyone has seen one.

Author: Sure. I just thought of something slightly less trivial, but you're right, let's keep it simple.

  WellNotifier(int availableToDrink) : super(availableToDrink);

  void drink(int liters) => state = min(state - liters, 0);
}

final wellNotifierProvider = StateNotifierProvider<WellNotifier, int>((ref) {

>   final availableToDrink = ref.watch(availableWater);

Reviewer: ```suggestion
  final availableWater = ref.watch(availableWaterProvider);
```

  return WellNotifier(availableToDrink);
});
```

Let's recap the main differences:

- `StateNotifier`'s reactive dependencies are declared in its provider, whereas `Notifier`
  centralizes this logic in its `build` method

> - Indeed, `StateNotifier`'s whole initialization process is split between its provider and its constructor,

Reviewer: ```suggestion
- `StateNotifier`'s whole initialization process is split between its provider and its constructor,
```

  whereas `Notifier` reserves a single place to place such logic
- Notice how, as opposed to `StateNotifier`, no logic is ever written into a `Notifier`'s constructor

Similar conclusions can be made with `AsyncNotifer`, `Notifier`'s asynchronous equivalent.  
Let's check it out.

## Better async support with `AsyncNotifier`

The main appeal of the new API syntax is an improved DX on asynchronous states.  
Previously, with `StateNotifier`, we would handle that like so:

```dart
class AsyncTodosNotifier extends StateNotifier<AsyncValue<List<Todo>>> {
  AsyncTodosNotifier() : super(const AsyncLoading()) {
    _postInit();
  }

  Future<void> _postInit() async {
    state = await AsyncValue.guard(
      // mock of a network request
      () => Future.delayed(const Duration(seconds: 1), () => []),
    );
  }

  // ...
}
```

> This heavy syntax suffers from the same defects showed above.  

Reviewer: What defect?

`AsyncNotifer`, just like `Notifier`, brings a simple and uniform API. Here, it's easy to see
`AsyncNotifer` as a `FutureProvider` with methods.

Here's the above example, rewritten with the new `AsyncNotifier` APIs:

<AutoSnippet language="dart" {...asyncNotifier}></AutoSnippet>

> Therefore, migrating from `StateNotifier<AsyncValue<T>>` to a `AsyncNotifer<T>` should be very straightforward:

Reviewer: ```suggestion
Therefore, migrating from `StateNotifier<AsyncValue<T>>` to a `AsyncNotifer<T>` should be a matter of:
```

- Place the initialization logic directly into `build`: there's no need to define a `_postInit` method anymore

> - There's no need to explicitly initialize `state` with `AsyncLoading`: as long as `build` returns a `Future`,

Reviewer: That "as long" bit isn't very clear to me

  this will be handled by `AsyncNotifer`, directly
- You can evaluate an initialization synchronously, if ever needed: just avoid returning a `Future` in your `build` method,
  i.e. let `build` be synchronous (this is the equivalent of initializing `StateNotifier` with an `AsyncData`)

Finally, `AsyncNotifer` comes with a set of utilities and getters that `StateNotifier` doesn't have, such as e.g.
[`future`](https://pub.dev/documentation/riverpod/latest/riverpod/AutoDisposeAsyncNotifier/future.html)
and [`update`](https://pub.dev/documentation/riverpod/latest/riverpod/AutoDisposeAsyncNotifier/update.html).

> This enables us to write much simpler logic when handling asynchronous mutations.

Reviewer: I'd rather redirect folks to the performing side effect page

## Explicit `.family` and `.autoDispose` modifications

> When a `StateNotifier` had family parameters, these were not found into its class, but they were

Reviewer: It still is today ;) 
```suggestion
When a `StateNotifier` has family parameters, these are not found into its class, but rather passed to its provider
```

> When a `StateNotifier` had family parameters, these were not found into its class, but they were

Reviewer: I'd drop this first sentence entirely and focus uniquely on how notifiers have a different subclass based on the modifiers.

Reviewer: We could have a small `:::info` later down talking about how family args are available with `this.argument` (or this.myVariable with codegen)

defined in its provider; also, and chances are that these parameters must have been piped down its constructor, too.

> `Notifier`, instead, has its own `.family` and `.autoDispose` counterparts, such as `FamilyNotifier`

Reviewer: This feels unrelated to the previous paragraph.

Author: Indeed it isn't. I'll try to introduce it differently

  BugsEncounteredNotifier({
    required this.ref,
    required this.featureId,
  }) : super(const AsyncData(99));
  final String featureId;
  final Ref ref;

  Future<void> fix(int amount) async {
    state = await AsyncValue.guard(() async {
      final old = state.requireValue;
      final result = await ref.read(taskTrackerProvider).fix(id: featureId, fixed: amount);
      return max(old - result, 0);
    });
  }
}

final bugsEncounteredNotifierProvider =
    StateNotifierProvider.family.autoDispose<BugsEncounteredNotifier, int, String>((ref, id) {
  return BugsEncounteredNotifier(ref: ref, featureId: id);
});
```

The main gotcha here is that reading `BugsEncounteredNotifier` is _not_ enough to determine wheter or
not this `StateNotifier` has modifications (e.g. `.family` and/or `.autoDispose`).  
To determine that, it is required to read the its provider definition.

Now, let's take a look at its migrated `AsyncNotifier` counterpart:

<AutoSnippet language="dart" {...familyAndDispose}></AutoSnippet>

This last snippet should showcase better readability: it's possible to immediately tell
what modifications are being set for this Notifier.

> ## Lifecycles have a different behavior

Reviewer: We should talk about a dispose override vs onDispose

Lifecycles between `Notifier`/`AsyncNotifier` and `StateNotifier`/`ChangeNotifier` differ substantially.

> `StateNotifier` and `ChangeNotifier` instances are indeed bound to their outer

Reviewer: I'm considering making a breaking change to change this in Riverpod 3. To have a `mounted`.

`StateNotifierProvider` and `ChangeNotifierProvider` providers.  
When their provider rebuilds, their exposed value, aka their `StateNotifier`/`ChangeNotifier` instance,
is `disposed`, and then re-created (i.e. its constructor gets invoked again).

> This example showcases - again - how the old API have sparse logic:

Reviewer: I don't think that's something worth mentioning. To begin with, there's not a big difference with the new approach here anyway.

```dart
class MyNotifier extends StateNotifier<int> {
  MyNotifier(this.period) : super(0) { // 1 init logic
    _timer = Timer.periodic(period, (t) => update());  // 2 side effect on init
  }
  final Duration period;
  late final Timer _timer;

  void update() => state++;  // 3 mutation

  @override
  void dispose() {
    _timer.cancel();  // 4 custom dispose logic
    super.dispose();
  }
}

final myNotifierProvider = StateNotifierProvider<MyNotifier, int>((ref) {  // 5 provider definition
  final Duration period = ref.watch(durationProvider);   // 6 reactive dependency logic
  return MyNotifier(period);
});
```

Here, if `durationProvider` updates, `MyNotifier` _disposes_: its instance is then re-instantiated
and its internal state is then re-initialized.  
Furthermore, unlike every other provider, the `dispose` callback is to be defined
in the class, separately.  

> Finally, it is still possible to write `ref.onDispose` in its _provider_, showing once again how

Reviewer: That should be a full-on section talking about the migration for this.

  MyNotifier() : super(0);
}

final myNotifierProvider = StateNotifierProvider<MyNotifier, int>((ref) {
  final notifier = MyNotifier();

  final cleanup = notifier.addListener((state) => debugPrint('$state'));
  ref.onDispose(cleanup);

  // Or, equivalently:
  // final listener = notifier.stream.listen((event) => debugPrint('$event'));
  // ref.onDispose(listener.cancel);

  return notifier;
});
```
Becomes this:
<AutoSnippet language="dart" {...addListener}></AutoSnippet>

In a nutshell: if you want to listen to a `Notifier`/`AsyncNotifer`, just use `ref.listen`,
[as explained here](/docs/concepts/reading#using-reflisten-to-react-to-a-provider-change).

### From `.debugState`

`StateNotifier` exposed `.debugState`: this property had some use within `state_notifier`,
as it enabled accessing state from outside the class when in development mode, for testing purposes.

Again, for uniformity reasons, such property is not available in the new APIs.  
If you want to interact with `Notifier`/`AsyncNotifier` in a testing environment, you should do
it just like with any other provider: by reading them directly (as opposed of directly instantiating

> a `Notifier` by hand).

Reviewer: Not sure what you mean.

`Notifier.state` is accessible by tests if they wanted to. It's `@visibleForTesting`.

Author: I'm trying to write down your bullet points in #2630.

There, I read:
> StateNotifier.debugState no longer exists.
Tests which want to interact with Notifiers should always do so by reading the provider instead of directly instantiating the Notifier by hand.
You can do providerContainer.read(provider) to get the current state of a provider inside tests.

So, what do I do with this? I can delete the whole paragraph if you want me to.

Reviewer: Ah right. Those bullet points aren't quite organized. It's mix and matching a few things

I'd split it in two. There are two things going on:
- Notifiers shouldn't be instantiated by hand. They should be interacted with by using the provider
  Failing to do so will break the notifier due to ref & family args not being initialized
- StateNotifier.debugState doesn't exist. If you have a notifier, you can access `.state` inside tests freely.
  If you don't have the Notifier instance, read the provider to obtain one. Cf above

While this example has been crafted to show how `ChangeNotifier` can lead to faulty design choices
for newbie developers, the main takeaway is that mutable state might be harder than it looks.

`Notifier`/`AsyncNotifer`, in combination with immutable state, can lead to better design choices
and less errors; this approach lifts the developer from several intricacies due to the nature
of asynchronous state.

With `AsyncNotifier`, the above becomes:
<AutoSnippet language="dart" {...fromChangeNotifier}></AutoSnippet>

### Migration Process
Let's analyze the migration process applied here:
1. We've moved the initialization from a method invoked in a constructor, directly into `build`
2. We've removed `todos`, `isLoading` and `hasError`: `state` suffice
3. Notice how, since `build` returns a `Future<T>`, state is implicitly initialized with `AsyncLoading`,
waiting for the initial request to finish
4. We've returned the result of the (mocked) asynchronous request, with no try-catch-finally blocks
5. We've then simplified `addTodo`, exploiting `AsyncValue`'s APIs
6. Notice that, to apply mutations, we simply reassign `state` directly:

> listeners will be automatically notified about it

Reviewer: It feels weird that in a document about migration, the migration bit is only 10 lines long.

> 

Reviewer: We don't have a single showcase of ChangeNotifier migration.

Reviewer: Wait my bad, I missed it

### Advantages
Finally, let's highlight the main advantages of the new APIs:
- There's a lot less code, and what's left is way more simple and readable
- Since `AsyncNotifier` implicitly uses `AsyncValue`, there's no need to manually define and handle
 `isLoading` and `hasError`
- There's no need to explicitly handle errors (i.e. try-catch-finally blocks), since
`AsyncNotifier.build` converts errors into `AsyncError` and valid data into `AsyncData`

> - `AsyncValue.guard` essentialy emulates `AsyncNotifier.build`, greatly simplifying mutations

Reviewer: This feels redundant and way too far in the page. That should be in the explanation of the syntax
